src/main/java/com/squareup/javapoet/ArrayTypeName.java

```diff
@@ -43,6 +43,26 @@ private ArrayTypeName(TypeName componentType, List<AnnotationSpec> annotations)
     return new ArrayTypeName(componentType, concatAnnotations(annotations));
   }
 
+  @Override StringBuilder binary(StringBuilder stringBuilder) {
+    stringBuilder.append('[');
+    if (componentType.isPrimitive()) {
+      if (componentType.equals(TypeName.BOOLEAN)) return stringBuilder.append('Z');
+      if (componentType.equals(TypeName.BYTE)) return stringBuilder.append('B');
+      if (componentType.equals(TypeName.CHAR)) return stringBuilder.append('C');
+      if (componentType.equals(TypeName.DOUBLE)) return stringBuilder.append('D');
+      if (componentType.equals(TypeName.FLOAT)) return stringBuilder.append('F');
+      if (componentType.equals(TypeName.INT)) return stringBuilder.append('I');
+      if (componentType.equals(TypeName.LONG)) return stringBuilder.append('J');
+      if (componentType.equals(TypeName.SHORT)) return stringBuilder.append('S');
+      throw new IllegalStateException(componentType + " is primitive and not handled?!");
```

Comment:
Typically this is better as a naked `AssertionError`. We don’t want to waste bytecodes or sourcecode building strings that’ll never get built.